src/inmanta/execute/scheduler.py

```diff
@@ -430,3 +470,220 @@ def run(self, compiler: "Compiler", statements: Sequence["Statement"], blocks: S
+        }
+        self._tmp_list_variables = deque()
+        for drv in drvs:
+            self.append(drv, dont_reset_working_list=True)
+        self._freeze_order_working_list = self._freeze_order.copy()
+
+
+class CycleInTypeHintsError(Exception):
+    """
+    Raised when a cycle exists in the type hints provided to the compiler.
+    """
+
+    pass
+
+
+class TypePrecedenceGraph:
+    """
+    A graph representation of the type hints provided to the compiler.
+    """
+
+    def __init__(self, type_hints: List["TypeHint"] = []) -> None:
+        # The root nodes of the graph, where all other nodes attach to.
+        self.root_nodes: Set[TypePrecedenceGraphNode] = set()
+        # Dict of all nodes in the graph
+        self.type_to_node: Dict[EntityRelationship, TypePrecedenceGraphNode] = {}
+        # Creates nodes in graph
+        for hint in type_hints:
+            first_type, then_type = EntityRelationship.from_type_hint(hint)
+            self.add_precedence_rule(first_type, then_type)
+
+    def add_precedence_rule(self, first_type: EntityRelationship, then_type: EntityRelationship) -> None:
+        """
+        Add a rule that `first_type` should be frozen before `then_type`.
+        """
+        first_node = self._get_or_create_node(first_type, attach_to_root=True)
+        then_node = self._get_or_create_node(then_type, attach_to_root=False)
+        first_node.add_dependent(then_node)
+
+    def _get_or_create_node(self, entity_relationship: EntityRelationship, attach_to_root: bool) -> "TypePrecedenceGraphNode":
+        """
+        Get the TypePrecedenceGraphNode for the given entity_relationship in this graph
+        or create a new node if no such node exists.
+
+        :param attach_to_root: Whether a newly created node should be added to the
+                               root_nodes.
+        """
+        if entity_relationship not in self.type_to_node:
+            node = TypePrecedenceGraphNode(entity_relationship)
+            self.type_to_node[entity_relationship] = node
+            if attach_to_root:
+                self.root_nodes.add(node)
+            return node
+        else:
+            return self.type_to_node[entity_relationship]
+
+    def get_freeze_order(self) -> List[EntityRelationship]:
+        """
+        Return all the EntityRelationship in this graph in the order in which
+        they should be frozen.
+        """
+        if not self.root_nodes:
+            return []
+        work: Set[TypePrecedenceGraphNode] = set(self.root_nodes)
+        result: List[EntityRelationship] = []
+
+        def get_next_ready_item_in_work() -> TypePrecedenceGraphNode:
+            assert work
+            for current_node in work:
+                if all(dep.entity_relationship in result for dep in current_node.dependencies):
+                    return current_node
+            raise CycleInTypeHintsError("Cycle in type hints")
```

Comment:
Should I print the set of rules that cause the cycle. I assume that's not required, since the set of type will should be rather limited at all times. Correct?